src/node/rpc/handlerregistry.h

```diff
@@ -0,0 +1,241 @@
+      Forwardable forwardable = Forwardable::CanForward,
+      bool execute_locally = false)
+    {
+      handlers[method] = {
+        f, rw, params_schema, result_schema, forwardable, execute_locally};
+    }
+
+    void install(
+      const std::string& method,
+      HandleFunction f,
+      ReadWrite rw,
+      Forwardable forwardable)
+    {
+      install(
+        method,
+        f,
+        rw,
+        nlohmann::json::object(),
+        nlohmann::json::object(),
+        forwardable);
+    }
+
+    template <typename In, typename Out, typename F>
+    void install_with_auto_schema(
+      const std::string& method,
+      F&& f,
+      ReadWrite rw,
+      Forwardable forwardable = Forwardable::CanForward,
+      bool execute_locally = false)
+    {
+      auto params_schema = nlohmann::json::object();
+      if constexpr (!std::is_same_v<In, void>)
+      {
+        params_schema = ds::json::build_schema<In>(method + "/params");
+      }
+
+      auto result_schema = nlohmann::json::object();
+      if constexpr (!std::is_same_v<Out, void>)
+      {
+        result_schema = ds::json::build_schema<Out>(method + "/result");
+      }
+
+      install(
+        method,
+        std::forward<F>(f),
+        rw,
+        params_schema,
+        result_schema,
+        forwardable,
+        execute_locally);
+    }
+
+    template <typename T, typename... Ts>
+    void install_with_auto_schema(const std::string& method, Ts&&... ts)
+    {
+      install_with_auto_schema<typename T::In, typename T::Out>(
+        method, std::forward<Ts>(ts)...);
+    }
+
+    /** Set a default HandleFunction
+     *
+     * The default HandleFunction is only invoked if no specific HandleFunction
+     * was found.
+     *
+     * @param f Method implementation
+     * @param rw Flag if method will Read, Write, MayWrite
+     */
+    void set_default(HandleFunction f, ReadWrite rw)
+    {
+      default_handler = {f, rw};
+    }
+
+    /** Populate out with all supported methods
+     *
+     * This is virtual since the default handler may do its own dispatch
+     * internally, so derived implementations must be able to populate the list
+     * with the supported methods however it constructs them.
+     */
+    virtual void list_methods(Store::Tx& tx, ListMethods::Out& out)
+    {
+      for (const auto& handler : handlers)
+      {
+        out.methods.push_back(handler.first);
+      }
+    }
+
+    virtual void init_handlers(Store& tables) {}
+
+    virtual Handler* find_handler(const std::string& method)
+    {
+      auto search = handlers.find(method);
+      if (search != handlers.end())
+      {
+        return &search->second;
+      }
+      else if (default_handler)
+      {
+        return &default_handler.value();
+      }
+
+      return nullptr;
+    }
+
+    virtual void tick(std::chrono::milliseconds elapsed, size_t tx_count) {}
+
+    virtual std::optional<CallerId> valid_caller(
```

Comment:
My first feeling is that this should belong to a different class as this feature is not related to the handlers installed? Or go back to `frontend.h`?

Reply:
I wrote a big blurb but didn't convince myself, so here's a short cop-out answer:

I'd like to pull as much out of `frontend.h` as possible, so that we have greater freedom to implement dispatch, auth, and validation as far down the stack as possible, in potentially handler-specific ways. In practice we don't have a use-case for handler-specific `valid_caller` yet, so this may make more sense back in `frontend.h` in future. But for now I'm happy to pull too many things out rather than too few, in the assumption that these will be tidied in a future pass.